gateway/tests/gateway_test.rs

```diff
@@ -675,40 +680,300 @@ fn test_send_cross() {
+                rt.expect_send(
+                    sto.clone(),
+                    METHOD_SEND,
+                    RawBytes::default(),
+                    value.clone(),
+                    RawBytes::default(),
+                    ExitCode::OK,
+                );
+            })),
+        )
+        .unwrap();
+    assert_eq!(cid.is_none(), true);
+}
+
+/// This test covers the case where a bottom up cross_msg's target subnet is NOT the same as that of
+/// the gateway. It would save in postbox.
+#[test]
+fn test_apply_msg_bu_not_target_subnet() {
+    // ============== Register subnet ==============
+    let shid = SubnetID::new_from_parent(&ROOTNET_ID, *SUBNET_ONE);
     let (h, mut rt) = setup(shid.clone());
 
     let from = Address::new_bls(&[3; fvm_shared::address::BLS_PUB_LEN]).unwrap();
     let to = Address::new_bls(&[4; fvm_shared::address::BLS_PUB_LEN]).unwrap();
 
-    let sub1 = SubnetID::new(&shid, *SUBNET_ONE);
-    let sub2 = SubnetID::new(&shid, *SUBNET_TWO);
+    let sub = shid.clone();
 
-    // register subnets
-    let reg_value = TokenAmount::from_atto(10_u64.pow(18));
-    h.register(&mut rt, &SUBNET_ONE, &reg_value, ExitCode::OK)
+    // ================ Setup ===============
+    let value = TokenAmount::from_atto(10_u64.pow(17));
+
+    // ================= Bottom-Up ===============
+    let ff = IPCAddress::new(&sub, &to).unwrap();
+    let tt = IPCAddress::new(&ROOTNET_ID, &from).unwrap();
+    let msg_nonce = 0;
+
+    // Only system code is allowed to this method
+    let params = StorableMsg {
+        to: tt.clone(),
+        from: ff.clone(),
+        method: METHOD_SEND,
+        value: value.clone(),
+        params: RawBytes::default(),
+        nonce: msg_nonce,
+    };
+    let cid = h
+        .apply_cross_execute_only(&mut rt, value.clone(), params, None)
+        .unwrap()
         .unwrap();
+
+    // Part 1: test the message is stored in postbox
+    let st: State = rt.get_state();
+    assert_ne!(tt.subnet().unwrap(), st.network_name);
+
+    // Check 1: `tt` is in `sub1`, which is not in that of `runtime` of gateway, will store in postbox
+    let item = st.load_from_postbox(rt.store(), cid.clone()).unwrap();
+    assert_eq!(item.owners, Some(vec![ff.clone().raw_addr().unwrap()]));
+    let msg = item.cross_msg.msg;
+    assert_eq!(msg.to, tt);
+    // the nonce should not have changed at all
+    assert_eq!(msg.nonce, msg_nonce);
+    assert_eq!(msg.value, value);
+
+    // Part 2: Now we propagate from postbox
+    // get the original subnet nonce first
+    let caller = ff.clone().raw_addr().unwrap();
+    h.propagate(&mut rt, caller, cid.clone()).unwrap();
+
+    // state should be updated, load again
+    let st: State = rt.get_state();
+
+    // cid should be removed from postbox
+    let r = st.load_from_postbox(rt.store(), cid.clone());
+    assert_eq!(r.is_err(), true);
+    let err = r.unwrap_err();
+    assert_eq!(err.to_string(), "cid not found in postbox");
```

Comment:
Can we check here also that after propagate the message has been committed correctly and that `commit_bottom_up` was triggered correctly?

Reply:
The nonce can be checked easily which I added, but the checkpoint mechanism, I still need to understand first... I actually not sure what's happening inside those chunk yet, haha...